### ISSUE TITLE
ptset/ptmap are broken under 4.05 (beta)

### ISSUE THREAD

**Issue by gasche** [backtracking/ptset#1]
opam-builder reports:
  http://opam.ocamlpro.com/builder/html/ptset/ptset.1.0.0/2df016054683ba0af52a1ddae1f33d8a
  http://opam.ocamlpro.com/builder/html/ptmap/ptmap.2.0.1/37f00ce2a1b7c7371cfab67b79c734aa

> Error: The implementation ptset.ml
> does not match the interface dist/build/lib-ptset/opt-d/ptset.cmi:
>
>        The value `find_last_opt' is required but not provided
>        The value `find_last' is required but not provided
>        The value `find_first_opt' is required but not provided
>        The value `find_first' is required but not provided
>        The value `find_opt' is required but not provided
>        The value `choose_opt' is required but not provided
>        The value `max_elt_opt' is required but not provided
>        The value `min_elt_opt' is required but not provided

> File "ptmap.ml", line 1:
> Error: The implementation ptmap.ml
> does not match the interface dist/build/lib-ptmap/ptmap.cmi:
>
>        The value `find_last_opt' is required but not provided
>        The value `find_last' is required but not provided
>        The value `find_first_opt' is required but not provided
>        The value `find_first' is required but not provided
>        The value `choose_opt' is required but not provided
>        The value `max_binding_opt' is required but not provided
>        The value `min_binding_opt' is required but not provided

I suppose that you can either stop using the standard `Set.S` interface (by duplicating it) to not break on new releases, or implement the new functions. They come from https://github.com/ocaml/ocaml/pull/869 .

**Comment by UnixJunkie** [comment]
Duplicating anything violates a software engineering principle.

**Comment by UnixJunkie** [comment]
I will use cppo to handle ocaml versions >= 4.05.

**Comment by gasche** [comment]
I don't think you need to use a preprocessor: why not implement these functions for all OCaml versions? At worst, they will be hidden by a more restrictive signature on older versions.

(In Batteries we do need a preprocessor because some of the features we expose depend on new runtime features that cannot be re-implemented externally. This is not the case here.)

**Comment by UnixJunkie** [comment]
Good idea, I will try.

**Comment by UnixJunkie** [comment]
related to https://github.com/backtracking/ptset/pull/2

**Comment by backtracking** [comment]
I prefer the idea of abandoning the compatibility with Set.S, in particular because functions such as
min_elt, max_elt, or elements make no sense with Patricia trees.
I dislike the idea of duplicating a fragment of Set.S, but I prefer this solution anyway.
